### PR TITLE
Service access control

### PR DESCRIPTION
**Pull request by mb-dbc-dk**
Check remote ip and authentication triple

### CODE REVIEW

File: service/src/main/java/dk/dbc/rr/oai/OaiBean.java [hunk 1]
```diff
@@ -40,14 +50,66 @@
...
+                    .entity(indexHtml.getInputStream())
+                    .build();
+        }
+
+        String triple = params.getFirst("identity");
+        if (triple == null || triple.split(":", 3).length != 3)
+            triple = headers.getHeaderString("Identity");
+        String clientIp = remoteIp.clientIp(httpRequest.getRemoteAddr(),
+                                            headers.getHeaderString("X-Forwarded-For"));
+        log.trace("tripplet = {}; clientIp = {}", triple, clientIp);
```
Comment (noah-dbc) [inline service/src/main/java/dk/dbc/rr/oai/OaiBean.java]: spelling of "tripplet"

File: service/src/main/java/dk/dbc/rr/oai/OaiBean.java [hunk 2]
```diff
@@ -40,14 +50,66 @@
...
+
+    Set<String> getAllowedSets(String triple, String clientIp) throws ServerErrorException {
+        Set<String> allowedSets = EMPTY_SET;
+        if (config.isAuthenticationDisabled()) {
+            allowedSets = config.getAllForsRightsSets();
+        } else {
+            try {
+                allowedSets = forsRights.authorized(triple, clientIp);
+            } catch (IOException ex) {
+                log.error("Error validating user: {}", ex.getMessage());
```
Comment (noah-dbc) [inline service/src/main/java/dk/dbc/rr/oai/OaiBean.java]: "validating user"? Not saying it's not what is happening here, but I think the message here could be clearer.

File: service/src/main/java/dk/dbc/rr/oai/fetch/forsrights/ForsRights.java [hunk 3]
```diff
@@ -69,16 +70,20 @@
...
             log.info("Fetching forsrights from: {}", uri.toString().replaceFirst("&passwordAut=[^&]*", "&passwordAut=[REDACTED]"));
             DTO dto = getDTO(uri);
             log.debug("dto = {}", dto);
-            return config.getForsRightsRules().entrySet().stream()
-                    .filter(e -> dto.hasRight(e.getKey())) // Map entries that are allowd by DTO
-                    .flatMap(e -> e.getValue().stream())
-                    .collect(toSet());
+            if (dto.hasAnyRight()) {
+                return config.getForsRightsRules().entrySet().stream()
+                        .filter(e -> dto.hasRight(e.getKey())) // Map entries that are allowd by DTO
```
Comment (noah-dbc) [inline service/src/main/java/dk/dbc/rr/oai/fetch/forsrights/ForsRights.java]: typo in comment

File: service/src/test/java/dk/dbc/rr/oai/OaiBeanIT.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        System.out.println("testGetAllowedSets");
+        Set<String> unknown = oaiBean.getAllowedSets(UNAUTHORIZED_TRIPPLET, "127.0.0.1");
+        assertThat(unknown.stream().sorted().collect(toList()), is(asList()));
+        Set<String> an = oaiBean.getAllowedSets(AN_TRIPLET, "127.0.0.1");
+        assertThat(an.stream().sorted().collect(toList()), is(asList("art", "nat")));
+        Set<String> abno = oaiBean.getAllowedSets("abno:xxx:yyy", "127.0.0.1");
+        assertThat(abno.stream().sorted().collect(toList()), is(asList("art", "bkm", "nat", "onl")));
+    }
+    private static final String AN_TRIPLET = "an:xxx:yyy";
+    private static final String UNAUTHORIZED_TRIPPLET = "unknown:xxx:yyy";
```
Comment (noah-dbc) [inline service/src/test/java/dk/dbc/rr/oai/OaiBeanIT.java]: two Ps

File: service/src/main/java/dk/dbc/rr/oai/OaiBean.java (hunk 2 above)
Comment (mb-dbc-dk) [reply service/src/main/java/dk/dbc/rr/oai/OaiBean.java]: This is only if a user validation has technical problems ie. the server is unreachable. If the call is successful but the user isn't validated allowed sets will be empty. Given the design of ForsRights, we cannot tell if the the user didn't validate (password) or the rights to this user wasn't granted (permission).

Comment (mb-dbc-dk) [reply service/src/main/java/dk/dbc/rr/oai/OaiBean.java]: Error communicating with ForsRights

**Pull request merged**